File: src/lib/coachPlatform.ts
export const COACH_CONSENT_SCOPES = [
  'nutrition',
  'workouts',
  'activity',
  'hydration',
  'supplements',
  'avatar',
  'measurements',
  'notes',
  'recovery',
  'visual_progress',
] as const

export type CoachConsentScope = (typeof COACH_CONSENT_SCOPES)[number]
export type CoachRelationshipStatus = 'invited' | 'active' | 'grace' | 'ended'
export type CoachSeatState = 'pending' | 'active' | 'grace' | 'released'

export interface CoachProfileSummary {
  status: 'development' | 'active' | 'suspended'
  display_name: string
  seat_limit: number
  active_seats: number
}

export interface CoachSponsorshipSummary {
  relationship_id: string
  coach_display_name: string
  relationship_status: CoachRelationshipStatus
  seat_state: CoachSeatState
  offered_scopes: CoachConsentScope[]
  consented_scopes: CoachConsentScope[]
  grace_ends_at: string | null
}

export interface CoachClientPolicyInput {
  relationship_status: CoachRelationshipStatus | null
  seat_state: CoachSeatState | null
  consented_scopes: readonly CoachConsentScope[]
  individual_access: boolean
}

export interface CoachClientPolicy {
  can_use_sponsored_app: boolean
  can_follow_coach_plan: boolean
  coach_plan_read_only: boolean
  can_create_custom_workouts: boolean
  can_rebuild_fitness_plan: boolean
  can_use_orbit: boolean
  can_use_nutrition: boolean
  can_use_avatar: boolean
  can_view_visual_progress: boolean
}

export function coachClientPolicy(input: CoachClientPolicyInput): CoachClientPolicy {
  const activeSponsorship = input.relationship_status === 'active' && input.seat_state === 'active'
  const grace = input.relationship_status === 'grace' || input.seat_state === 'grace'
  const relationshipExists = activeSponsorship || grace
  return {
    can_use_sponsored_app: activeSponsorship,
    can_follow_coach_plan: activeSponsorship,
    coach_plan_read_only: grace,
    can_create_custom_workouts: input.individual_access || !relationshipExists,
    can_rebuild_fitness_plan: input.individual_access || !relationshipExists,
    can_use_orbit: input.individual_access || !relationshipExists,
    can_use_nutrition: input.individual_access || !relationshipExists || activeSponsorship,
    can_use_avatar: input.individual_access || !relationshipExists || activeSponsorship,
    can_view_visual_progress: input.individual_access || !relationshipExists,
  }
}

export type CoachPlanChecklistKey =
  | 'nutrition'
  | 'workouts'
  | 'supplements'
  | 'hydration'
  | 'schedule'
  | 'review_date'

export interface CoachPlanChecklist extends Record<CoachPlanChecklistKey, boolean> {}

export type CoachExerciseUnit = 'reps' | 'seconds' | 'minutes' | 'metres' | 'steps' | 'rounds'

export interface CoachExerciseTemplate {
  id: string
  movement_id: string
  name: string
  sets: number
  target_min: number
  target_max: number
  unit: CoachExerciseUnit
  per_side: boolean
  rest_seconds: number
  tempo_up_seconds: number
  tempo_down_seconds: number
  tempo_pause_seconds: number
  notes: string
  optional: boolean
  group_id: string | null
  group_position: number | null
}

export interface CoachSessionTemplate {
  id: string
  weekday: number
  name: string
  session_mode: 'guided' | 'tracked'
  estimated_minutes: number
  warmup_note: string
  exercises: CoachExerciseTemplate[]
}

export interface CoachPlanDraft {
  title: string
  objective: string
  coach_note: string
  review_date: string | null
  checklist: CoachPlanChecklist
  sessions: CoachSessionTemplate[]
}

export interface CoachCurrentPlan {
  id: string
  relationship_id: string
  version: number
  status: 'draft' | 'published' | 'superseded'
  title: string
  objective: string
  coach_note: string
  review_date: string | null
  checklist: CoachPlanChecklist
  plan: CoachPlanDraft
  published_at: string | null
  acknowledged_at: string | null
  activated_at: string | null
}

export interface CoachAccountContext {
  coach: CoachProfileSummary | null
  sponsorship: CoachSponsorshipSummary | null
  current_plan: CoachCurrentPlan | null
  capabilities: {
    coach_workspace: boolean
    sponsored_client: boolean
  }
}

export const EMPTY_COACH_ACCOUNT_CONTEXT: CoachAccountContext = {
  coach: null,
  sponsorship: null,
  current_plan: null,
  capabilities: { coach_workspace: false, sponsored_client: false },
}

export interface CoachRosterEntry {
  id: string
  client_user_id: string
  display_name: string
  relationship_status: CoachRelationshipStatus
  seat_state: CoachSeatState
  consented_scopes: CoachConsentScope[]
  plan_version: number | null
  plan_title: string | null
  review_date: string | null
  published_at: string | null
  acknowledged_at: string | null
  activated_at: string | null
  attention: CoachRosterAttention[]
}

export interface CoachClientOverview {
  relationship_id: string
  client_user_id: string
  display_name: string
  relationship_status: CoachRelationshipStatus
  seat_state: CoachSeatState
  consented_scopes: CoachConsentScope[]
  measurements: null | {
    sex: string
    height_cm: number
    weight_kg: number
    body_fat_pct: number | null
    birthdate: string
  }
  avatar: null | Record<string, string | number | null>
  workouts: null | { completed_30d: number; last_completed_at: string | null }
  nutrition: null | { days_observed: number; average_kcal: number | null }
  hydration: null | { days_observed: number; average_litres: number | null }
  visual_progress_shared: boolean
  current_plan: CoachCurrentPlan | null
}

export interface CoachInvitationReceipt {
  invitation_id: string
  token: string
  expires_at: string
}

export interface CoachInvitationPreview {
  coach_display_name: string
  requested_scopes: CoachConsentScope[]
  visual_progress_requested: boolean
  expires_at: string
}

export interface CoachPlanVersionReceipt {
  id: string
  relationship_id: string
  version: number
  status: 'draft' | 'published' | 'superseded'
}

export interface CoachPlanIssue {
  code:
    | 'unknown_field'
    | 'too_long'
    | 'title_required'
    | 'objective_required'
    | 'review_date'
    | 'checklist'
    | 'sessions_required'
    | 'session_count'
    | 'duplicate_id'
    | 'weekday'
    | 'session_name'
    | 'session_mode'
    | 'estimated_minutes'
    | 'exercise_count'
    | 'movement'
    | 'exercise_name'
    | 'sets'
    | 'target'
    | 'target_order'
    | 'unit'
    | 'rest'
    | 'tempo'
    | 'group'
  path: string
}

const PLAN_FIELDS = new Set(['title', 'objective', 'coach_note', 'review_date', 'checklist', 'sessions'])
const CHECKLIST_FIELDS = new Set<CoachPlanChecklistKey>([
  'nutrition', 'workouts', 'supplements', 'hydration', 'schedule', 'review_date',
])
const SESSION_FIELDS = new Set([
  'id', 'weekday', 'name', 'session_mode', 'estimated_minutes', 'warmup_note', 'exercises',
])
const EXERCISE_FIELDS = new Set([
  'id', 'movement_id', 'name', 'sets', 'target_min', 'target_max', 'unit', 'per_side',
  'rest_seconds', 'tempo_up_seconds', 'tempo_down_seconds', 'tempo_pause_seconds',
  'notes', 'optional', 'group_id', 'group_position',
])
const EXERCISE_UNITS = new Set<CoachExerciseUnit>(['reps', 'seconds', 'minutes', 'metres', 'steps', 'rounds'])

function ownKeys(value: unknown): string[] {
  return value && typeof value === 'object' && !Array.isArray(value) ? Object.keys(value) : []
}

function unknownFields(value: unknown, allowed: ReadonlySet<string>, path: string): CoachPlanIssue[] {
  return ownKeys(value)
    .filter((key) => !allowed.has(key))
    .map((key) => ({ code: 'unknown_field', path: `${path}.${key}` }))
}

function boundedInteger(value: unknown, minimum: number, maximum: number): boolean {
  return Number.isInteger(value) && Number(value) >= minimum && Number(value) <= maximum
}

function validISODate(value: unknown): value is string {
  if (typeof value !== 'string' || !/^\d{4}-\d{2}-\d{2}$/.test(value)) return false
  const [year, month, day] = value.split('-').map(Number)
  const date = new Date(Date.UTC(year, month - 1, day))
  return date.getUTCFullYear() === year && date.getUTCMonth() === month - 1 && date.getUTCDate() === day
}

export function validateCoachPlan(
  value: CoachPlanDraft,
  options: { publishing: boolean; known_movement_ids?: ReadonlySet<string> },
): { publishable: boolean; issues: CoachPlanIssue[] } {
  const issues: CoachPlanIssue[] = unknownFields(value, PLAN_FIELDS, 'plan')
  const title = typeof value.title === 'string' ? value.title.trim() : ''
  const objective = typeof value.objective === 'string' ? value.objective.trim() : ''
  const coachNote = typeof value.coach_note === 'string' ? value.coach_note.trim() : ''
  if (title.length > 80) issues.push({ code: 'too_long', path: 'plan.title' })
  if (objective.length > 240) issues.push({ code: 'too_long', path: 'plan.objective' })
  if (coachNote.length > 4_000) issues.push({ code: 'too_long', path: 'plan.coach_note' })
  if (options.publishing && title.length < 2) issues.push({ code: 'title_required', path: 'plan.title' })
  if (options.publishing && objective.length < 2) issues.push({ code: 'objective_required', path: 'plan.objective' })
  if (value.review_date != null && !validISODate(value.review_date)) {
    issues.push({ code: 'review_date', path: 'plan.review_date' })
  }
  if (options.publishing && !validISODate(value.review_date)) {
    if (!issues.some((issue) => issue.code === 'review_date')) issues.push({ code: 'review_date', path: 'plan.review_date' })
  }

  issues.push(...unknownFields(value.checklist, CHECKLIST_FIELDS, 'plan.checklist'))
  if (options.publishing && (!value.checklist || [...CHECKLIST_FIELDS].some((key) => value.checklist[key] !== true))) {
    issues.push({ code: 'checklist', path: 'plan.checklist' })
  }

  const sessions = Array.isArray(value.sessions) ? value.sessions : []
  if (options.publishing && sessions.length === 0) issues.push({ code: 'sessions_required', path: 'plan.sessions' })
  if (sessions.length > 7) issues.push({ code: 'session_count', path: 'plan.sessions' })
  const identifiers = new Set<string>()
  sessions.forEach((session, sessionIndex) => {
    const path = `plan.sessions[${sessionIndex}]`
    issues.push(...unknownFields(session, SESSION_FIELDS, path))
    if (!session.id || identifiers.has(session.id)) issues.push({ code: 'duplicate_id', path: `${path}.id` })
    else identifiers.add(session.id)
    if (!boundedInteger(session.weekday, 1, 7)) issues.push({ code: 'weekday', path: `${path}.weekday` })
    if (typeof session.name !== 'string' || session.name.trim().length < 2 || session.name.trim().length > 80) {
      issues.push({ code: 'session_name', path: `${path}.name` })
    }
    if (session.session_mode !== 'guided' && session.session_mode !== 'tracked') {
      issues.push({ code: 'session_mode', path: `${path}.session_mode` })
    }
    if (!boundedInteger(session.estimated_minutes, 5, 360)) {
      issues.push({ code: 'estimated_minutes', path: `${path}.estimated_minutes` })
    }
    if (typeof session.warmup_note !== 'string' || session.warmup_note.length > 1_000) {
      issues.push({ code: 'too_long', path: `${path}.warmup_note` })
    }
    const exercises = Array.isArray(session.exercises) ? session.exercises : []
    if (options.publishing && exercises.length === 0) issues.push({ code: 'exercise_count', path: `${path}.exercises` })
    if (exercises.length > 30) issues.push({ code: 'exercise_count', path: `${path}.exercises` })
    exercises.forEach((exercise, exerciseIndex) => {
      const exercisePath = `${path}.exercises[${exerciseIndex}]`
      issues.push(...unknownFields(exercise, EXERCISE_FIELDS, exercisePath))
      if (!exercise.id || identifiers.has(exercise.id)) issues.push({ code: 'duplicate_id', path: `${exercisePath}.id` })
      else identifiers.add(exercise.id)
      if (
        typeof exercise.movement_id !== 'string' || exercise.movement_id.length < 2 ||
        (options.known_movement_ids && !options.known_movement_ids.has(exercise.movement_id))
      ) issues.push({ code: 'movement', path: `${exercisePath}.movement_id` })
      if (typeof exercise.name !== 'string' || exercise.name.trim().length < 2 || exercise.name.trim().length > 120) {
        issues.push({ code: 'exercise_name', path: `${exercisePath}.name` })
      }
      if (!boundedInteger(exercise.sets, 1, 12)) issues.push({ code: 'sets', path: `${exercisePath}.sets` })
      if (!boundedInteger(exercise.target_min, 1, 600) || !boundedInteger(exercise.target_max, 1, 600)) {
        issues.push({ code: 'target', path: `${exercisePath}.target` })
      } else if (exercise.target_min > exercise.target_max) {
        issues.push({ code: 'target_order', path: `${exercisePath}.target` })
      }
      if (!EXERCISE_UNITS.has(exercise.unit)) issues.push({ code: 'unit', path: `${exercisePath}.unit` })
      if (!boundedInteger(exercise.rest_seconds, 0, 600)) issues.push({ code: 'rest', path: `${exercisePath}.rest_seconds` })
      for (const key of ['tempo_up_seconds', 'tempo_down_seconds', 'tempo_pause_seconds'] as const) {
        const tempo = exercise[key]
        if (typeof tempo !== 'number' || !Number.isFinite(tempo) || tempo < 0 || tempo > 30) {
          issues.push({ code: 'tempo', path: `${exercisePath}.${key}` })
        }
      }
      if ((exercise.group_id == null) !== (exercise.group_position == null)) {
        issues.push({ code: 'group', path: `${exercisePath}.group` })
      }
      if (exercise.group_position != null && !boundedInteger(exercise.group_position, 1, 30)) {
        issues.push({ code: 'group', path: `${exercisePath}.group_position` })
      }
      if (typeof exercise.notes !== 'string' || exercise.notes.length > 1_000) {
        issues.push({ code: 'too_long', path: `${exercisePath}.notes` })
      }
    })
  })
  return { publishable: options.publishing && issues.length === 0, issues }
}

export type CoachRosterAttention =
  | 'plan_missing'
  | 'review_due'
  | 'awaiting_acknowledgement'
  | 'seat_grace'

export function coachRosterAttention(input: {
  relationship_status: CoachRelationshipStatus
  plan_version: number | null
  plan_published_at: string | null
  acknowledged_at: string | null
  review_date: string | null
  today: string
}): CoachRosterAttention[] {
  if (input.relationship_status === 'grace') return ['seat_grace']
  if (input.relationship_status !== 'active') return []
  if (input.plan_version == null) return ['plan_missing']
  const attention: CoachRosterAttention[] = []
  if (input.review_date && validISODate(input.review_date) && validISODate(input.today)) {
    const review = Date.parse(`${input.review_date}T00:00:00Z`)
    const today = Date.parse(`${input.today}T00:00:00Z`)
    if (review - today <= 7 * 86_400_000) attention.push('review_due')
  }
  if (input.plan_published_at && !input.acknowledged_at) attention.push('awaiting_acknowledgement')
  return attention
}

File: ios/APEXNative/APEX/Core/Engine/CoachPlatform.swift
import Foundation

enum CoachConsentScope: String, Codable, CaseIterable, Hashable, Sendable {
    case nutrition
    case workouts
    case activity
    case hydration
    case supplements
    case avatar
    case measurements
    case notes
    case recovery
    case visualProgress = "visual_progress"
}

enum CoachRelationshipStatus: String, Codable, Hashable, Sendable {
    case invited
    case active
    case grace
    case ended
}

enum CoachSeatState: String, Codable, Hashable, Sendable {
    case pending
    case active
    case grace
    case released
}

struct CoachClientPolicy: Equatable, Sendable {
    let canUseSponsoredApp: Bool
    let canFollowCoachPlan: Bool
    let coachPlanReadOnly: Bool
    let canCreateCustomWorkouts: Bool
    let canRebuildFitnessPlan: Bool
    let canUseOrbit: Bool
    let canUseNutrition: Bool
    let canUseAvatar: Bool
    let canViewVisualProgress: Bool

    static func resolve(
        relationshipStatus: CoachRelationshipStatus?,
        seatState: CoachSeatState?,
        consentedScopes: Set<CoachConsentScope>,
        individualAccess: Bool
    ) -> CoachClientPolicy {
        let active = relationshipStatus == .active && seatState == .active
        let grace = relationshipStatus == .grace || seatState == .grace
        let relationshipExists = active || grace
        return CoachClientPolicy(
            canUseSponsoredApp: active,
            canFollowCoachPlan: active,
            coachPlanReadOnly: grace,
            canCreateCustomWorkouts: individualAccess || !relationshipExists,
            canRebuildFitnessPlan: individualAccess || !relationshipExists,
            canUseOrbit: individualAccess || !relationshipExists,
            canUseNutrition: individualAccess || !relationshipExists || active,
            canUseAvatar: individualAccess || !relationshipExists || active,
            canViewVisualProgress: individualAccess || !relationshipExists
        )
    }
}

struct CoachPlanChecklist: Codable, Hashable, Sendable {
    var nutrition: Bool
    var workouts: Bool
    var supplements: Bool
    var hydration: Bool
    var schedule: Bool
    var reviewDate: Bool

    enum CodingKeys: String, CodingKey {
        case nutrition, workouts, supplements, hydration, schedule
        case reviewDate = "review_date"
    }

    static let empty = CoachPlanChecklist(
        nutrition: false,
        workouts: false,
        supplements: false,
        hydration: false,
        schedule: false,
        reviewDate: false
    )

    var isComplete: Bool {
        nutrition && workouts && supplements && hydration && schedule && reviewDate
    }
}

struct CoachExerciseTemplate: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var movementID: String
    var name: String
    var sets: Int
    var targetMin: Int
    var targetMax: Int
    var unit: String
    var perSide: Bool
    var restSeconds: Int
    var tempoUpSeconds: Double
    var tempoDownSeconds: Double
    var tempoPauseSeconds: Double
    var notes: String
    var optional: Bool
    var groupID: UUID?
    var groupPosition: Int?

    enum CodingKeys: String, CodingKey {
        case id, name, sets, unit, notes, optional
        case movementID = "movement_id"
        case targetMin = "target_min"
        case targetMax = "target_max"
        case perSide = "per_side"
        case restSeconds = "rest_seconds"
        case tempoUpSeconds = "tempo_up_seconds"
        case tempoDownSeconds = "tempo_down_seconds"
        case tempoPauseSeconds = "tempo_pause_seconds"
        case groupID = "group_id"
        case groupPosition = "group_position"
    }
}

struct CoachSessionTemplate: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var weekday: Int
    var name: String
    var sessionMode: WorkoutSessionMode
    var estimatedMinutes: Int
    var warmupNote: String
    var exercises: [CoachExerciseTemplate]

    enum CodingKeys: String, CodingKey {
        case id, weekday, name, exercises
        case sessionMode = "session_mode"
        case estimatedMinutes = "estimated_minutes"
        case warmupNote = "warmup_note"
    }
}

struct CoachPlanDraft: Codable, Hashable, Sendable {
    var title: String
    var objective: String
    var coachNote: String
    var reviewDate: String?
    var checklist: CoachPlanChecklist
    var sessions: [CoachSessionTemplate]

    enum CodingKeys: String, CodingKey {
        case title, objective, checklist, sessions
        case coachNote = "coach_note"
        case reviewDate = "review_date"
    }

    static let empty = CoachPlanDraft(
        title: "",
        objective: "",
        coachNote: "",
        reviewDate: nil,
        checklist: .empty,
        sessions: []
    )
}

struct CoachPlanValidationIssue: Equatable, Sendable {
    enum Code: String, Sendable {
        case tooLong
        case titleRequired
        case objectiveRequired
        case reviewDate
        case checklist
        case sessionsRequired
        case sessionCount
        case duplicateID
        case weekday
        case sessionName
        case estimatedMinutes
        case exerciseCount
        case movement
        case exerciseName
        case sets
        case target
        case targetOrder
        case unit
        case rest
        case tempo
        case group
    }

    let code: Code
    let path: String
}

enum CoachPlanValidator {
    struct Result: Equatable, Sendable {
        let publishable: Bool
        let issues: [CoachPlanValidationIssue]
    }

    private static let units: Set<String> = ["reps", "seconds", "minutes", "metres", "steps", "rounds"]

    static func validate(
        _ plan: CoachPlanDraft,
        publishing: Bool,
        knownMovementIDs: Set<String>? = nil
    ) -> Result {
        var issues: [CoachPlanValidationIssue] = []
        let title = plan.title.trimmingCharacters(in: .whitespacesAndNewlines)
        let objective = plan.objective.trimmingCharacters(in: .whitespacesAndNewlines)
        if title.count > 80 { issues.append(.init(code: .tooLong, path: "plan.title")) }
        if objective.count > 240 { issues.append(.init(code: .tooLong, path: "plan.objective")) }
        if plan.coachNote.count > 4_000 { issues.append(.init(code: .tooLong, path: "plan.coach_note")) }
        if publishing && title.count < 2 { issues.append(.init(code: .titleRequired, path: "plan.title")) }
        if publishing && objective.count < 2 { issues.append(.init(code: .objectiveRequired, path: "plan.objective")) }
        if let reviewDate = plan.reviewDate, validDate(reviewDate) == false {
            issues.append(.init(code: .reviewDate, path: "plan.review_date"))
        } else if publishing && plan.reviewDate == nil {
            issues.append(.init(code: .reviewDate, path: "plan.review_date"))
        }
        if publishing && !plan.checklist.isComplete {
            issues.append(.init(code: .checklist, path: "plan.checklist"))
        }
        if publishing && plan.sessions.isEmpty {
            issues.append(.init(code: .sessionsRequired, path: "plan.sessions"))
        }
        if plan.sessions.count > 7 { issues.append(.init(code: .sessionCount, path: "plan.sessions")) }

        var identifiers = Set<UUID>()
        for (sessionIndex, session) in plan.sessions.enumerated() {
            let path = String(format: "plan.sessions[%d]", sessionIndex)
            if !identifiers.insert(session.id).inserted {
                issues.append(.init(code: .duplicateID, path: "\(path).id"))
            }
            if !(1...7).contains(session.weekday) { issues.append(.init(code: .weekday, path: "\(path).weekday")) }
            let sessionName = session.name.trimmingCharacters(in: .whitespacesAndNewlines)
            if !(2...80).contains(sessionName.count) { issues.append(.init(code: .sessionName, path: "\(path).name")) }
            if !(5...360).contains(session.estimatedMinutes) {
                issues.append(.init(code: .estimatedMinutes, path: "\(path).estimated_minutes"))
            }
            if session.warmupNote.count > 1_000 { issues.append(.init(code: .tooLong, path: "\(path).warmup_note")) }
            if (publishing && session.exercises.isEmpty) || session.exercises.count > 30 {
                issues.append(.init(code: .exerciseCount, path: "\(path).exercises"))
            }
            for (exerciseIndex, exercise) in session.exercises.enumerated() {
                let exercisePath = String(format: "%@.exercises[%d]", path, exerciseIndex)
                if !identifiers.insert(exercise.id).inserted {
                    issues.append(.init(code: .duplicateID, path: "\(exercisePath).id"))
                }
                if exercise.movementID.count < 2 || (knownMovementIDs?.contains(exercise.movementID) == false) {
                    issues.append(.init(code: .movement, path: "\(exercisePath).movement_id"))
                }
                let exerciseName = exercise.name.trimmingCharacters(in: .whitespacesAndNewlines)
                if !(2...120).contains(exerciseName.count) {
                    issues.append(.init(code: .exerciseName, path: "\(exercisePath).name"))
                }
                if !(1...12).contains(exercise.sets) { issues.append(.init(code: .sets, path: "\(exercisePath).sets")) }
                if !(1...600).contains(exercise.targetMin) || !(1...600).contains(exercise.targetMax) {
                    issues.append(.init(code: .target, path: "\(exercisePath).target"))
                } else if exercise.targetMin > exercise.targetMax {
                    issues.append(.init(code: .targetOrder, path: "\(exercisePath).target"))
                }
                if !units.contains(exercise.unit) { issues.append(.init(code: .unit, path: "\(exercisePath).unit")) }
                if !(0...600).contains(exercise.restSeconds) { issues.append(.init(code: .rest, path: "\(exercisePath).rest_seconds")) }
                let tempo = [exercise.tempoUpSeconds, exercise.tempoDownSeconds, exercise.tempoPauseSeconds]
                if tempo.contains(where: { !$0.isFinite || $0 < 0 || $0 > 30 }) {
                    issues.append(.init(code: .tempo, path: "\(exercisePath).tempo"))
                }
                if (exercise.groupID == nil) != (exercise.groupPosition == nil)
                    || exercise.groupPosition.map({ !(1...30).contains($0) }) == true {
                    issues.append(.init(code: .group, path: "\(exercisePath).group"))
                }
                if exercise.notes.count > 1_000 { issues.append(.init(code: .tooLong, path: "\(exercisePath).notes")) }
            }
        }
        return Result(publishable: publishing && issues.isEmpty, issues: issues)
    }

    private static func validDate(_ value: String) -> Bool {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .iso8601)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.isLenient = false
        return formatter.date(from: value) != nil
    }
}

enum CoachProfileStatus: String, Codable, Hashable, Sendable {
    case development
    case active
    case suspended
}

struct CoachProfileSummary: Codable, Hashable, Sendable {
    let status: CoachProfileStatus
    let displayName: String
    let seatLimit: Int
    let activeSeats: Int

    enum CodingKeys: String, CodingKey {
        case status
        case displayName = "display_name"
        case seatLimit = "seat_limit"
        case activeSeats = "active_seats"
    }
}

struct CoachSponsorshipSummary: Codable, Hashable, Sendable {
    let relationshipID: UUID
    let coachDisplayName: String
    let relationshipStatus: CoachRelationshipStatus
    let seatState: CoachSeatState
    let offeredScopes: Set<CoachConsentScope>
    let consentedScopes: Set<CoachConsentScope>
    let graceEndsAt: String?

    enum CodingKeys: String, CodingKey {
        case relationshipID = "relationship_id"
        case coachDisplayName = "coach_display_name"
        case relationshipStatus = "relationship_status"
        case seatState = "seat_state"
        case offeredScopes = "offered_scopes"
        case consentedScopes = "consented_scopes"
        case graceEndsAt = "grace_ends_at"
    }
}

enum CoachPlanStatus: String, Codable, Hashable, Sendable {
    case draft
    case published
    case superseded
}

struct CoachCurrentPlan: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    let relationshipID: UUID
    let version: Int
    let status: CoachPlanStatus
    let title: String
    let objective: String
    let coachNote: String
    let reviewDate: String?
    let checklist: CoachPlanChecklist
    let plan: CoachPlanDraft
    let publishedAt: String?
    let acknowledgedAt: String?
    let activatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, version, status, title, objective, checklist, plan
        case relationshipID = "relationship_id"
        case coachNote = "coach_note"
        case reviewDate = "review_date"
        case publishedAt = "published_at"
        case acknowledgedAt = "acknowledged_at"
        case activatedAt = "activated_at"
    }
}

struct CoachAccountCapabilities: Codable, Hashable, Sendable {
    let coachWorkspace: Bool
    let sponsoredClient: Bool

    enum CodingKeys: String, CodingKey {
        case coachWorkspace = "coach_workspace"
        case sponsoredClient = "sponsored_client"
    }

    static let none = CoachAccountCapabilities(coachWorkspace: false, sponsoredClient: false)
}

struct CoachAccountContext: Codable, Hashable, Sendable {
    let coach: CoachProfileSummary?
    let sponsorship: CoachSponsorshipSummary?
    let currentPlan: CoachCurrentPlan?
    let capabilities: CoachAccountCapabilities

    enum CodingKeys: String, CodingKey {
        case coach, sponsorship, capabilities
        case currentPlan = "current_plan"
    }

    static let empty = CoachAccountContext(
        coach: nil,
        sponsorship: nil,
        currentPlan: nil,
        capabilities: .none
    )
}

enum CoachRosterAttention: String, Codable, Hashable, Sendable {
    case planMissing = "plan_missing"
    case reviewDue = "review_due"
    case awaitingAcknowledgement = "awaiting_acknowledgement"
    case seatGrace = "seat_grace"

    static func resolve(
        relationshipStatus: CoachRelationshipStatus,
        planVersion: Int?,
        planPublishedAt: String?,
        acknowledgedAt: String?,
        reviewDate: String?,
        today: String
    ) -> [CoachRosterAttention] {
        if relationshipStatus == .grace { return [.seatGrace] }
        guard relationshipStatus == .active else { return [] }
        guard planVersion != nil else { return [.planMissing] }
        var answer: [CoachRosterAttention] = []
        if let reviewDate,
           let review = ISO8601DateOnly.date(reviewDate),
           let anchor = ISO8601DateOnly.date(today),
           review.timeIntervalSince(anchor) <= 7 * 86_400 {
            answer.append(.reviewDue)
        }
        if planPublishedAt != nil && acknowledgedAt == nil { answer.append(.awaitingAcknowledgement) }
        return answer
    }
}

private enum ISO8601DateOnly {
    static func date(_ value: String) -> Date? {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .iso8601)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.isLenient = false
        return formatter.date(from: value)
    }
}

struct CoachRosterEntry: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    let clientUserID: UUID
    let displayName: String
    let relationshipStatus: CoachRelationshipStatus
    let seatState: CoachSeatState
    let consentedScopes: Set<CoachConsentScope>
    let planVersion: Int?
    let planTitle: String?
    let reviewDate: String?
    let publishedAt: String?
    let acknowledgedAt: String?
    let activatedAt: String?
    let attention: [CoachRosterAttention]

    enum CodingKeys: String, CodingKey {
        case id, attention
        case clientUserID = "client_user_id"
        case displayName = "display_name"
        case relationshipStatus = "relationship_status"
        case seatState = "seat_state"
        case consentedScopes = "consented_scopes"
        case planVersion = "plan_version"
        case planTitle = "plan_title"
        case reviewDate = "review_date"
        case publishedAt = "published_at"
        case acknowledgedAt = "acknowledged_at"
        case activatedAt = "activated_at"
    }
}

struct CoachInvitationReceipt: Codable, Hashable, Sendable {
    let invitationID: UUID
    let token: String
    let expiresAt: String

    enum CodingKeys: String, CodingKey {
        case token
        case invitationID = "invitation_id"
        case expiresAt = "expires_at"
    }
}

struct CoachInvitationPreview: Codable, Hashable, Sendable {
    let coachDisplayName: String
    let requestedScopes: Set<CoachConsentScope>
    let visualProgressRequested: Bool
    let expiresAt: String

    enum CodingKeys: String, CodingKey {
        case coachDisplayName = "coach_display_name"
        case requestedScopes = "requested_scopes"
        case visualProgressRequested = "visual_progress_requested"
        case expiresAt = "expires_at"
    }
}

struct CoachPlanVersionReceipt: Codable, Hashable, Sendable {
    let id: UUID
    let relationshipID: UUID
    let version: Int
    let status: CoachPlanStatus

    enum CodingKeys: String, CodingKey {
        case id, version, status
        case relationshipID = "relationship_id"
    }
}

struct CoachClientMeasurements: Codable, Hashable, Sendable {
    let sex: String
    let heightCM: Double
    let weightKG: Double
    let bodyFatPercent: Double?
    let birthdate: String

    enum CodingKeys: String, CodingKey {
        case sex, birthdate
        case heightCM = "height_cm"
        case weightKG = "weight_kg"
        case bodyFatPercent = "body_fat_pct"
    }
}

struct CoachClientWorkoutSummary: Codable, Hashable, Sendable {
    let completed30Days: Int
    let lastCompletedAt: String?

    enum CodingKeys: String, CodingKey {
        case completed30Days = "completed_30d"
        case lastCompletedAt = "last_completed_at"
    }
}

struct CoachClientNutritionSummary: Codable, Hashable, Sendable {
    let daysObserved: Int
    let averageKcal: Double?

    enum CodingKeys: String, CodingKey {
        case daysObserved = "days_observed"
        case averageKcal = "average_kcal"
    }
}

struct CoachClientHydrationSummary: Codable, Hashable, Sendable {
    let daysObserved: Int
    let averageLitres: Double?

    enum CodingKeys: String, CodingKey {
        case daysObserved = "days_observed"
        case averageLitres = "average_litres"
    }
}

struct CoachClientOverview: Codable, Hashable, Sendable {
    let relationshipID: UUID
    let clientUserID: UUID
    let displayName: String
    let relationshipStatus: CoachRelationshipStatus
    let seatState: CoachSeatState
    let consentedScopes: Set<CoachConsentScope>
    let measurements: CoachClientMeasurements?
    let avatar: [String: JSONValue]?
    let workouts: CoachClientWorkoutSummary?
    let nutrition: CoachClientNutritionSummary?
    let hydration: CoachClientHydrationSummary?
    let visualProgressShared: Bool
    let currentPlan: CoachCurrentPlan?

    enum CodingKeys: String, CodingKey {
        case measurements, avatar, workouts, nutrition, hydration
        case relationshipID = "relationship_id"
        case clientUserID = "client_user_id"
        case displayName = "display_name"
        case relationshipStatus = "relationship_status"
        case seatState = "seat_state"
        case consentedScopes = "consented_scopes"
        case visualProgressShared = "visual_progress_shared"
        case currentPlan = "current_plan"
    }
}

struct CoachPlanActivationReceipt: Codable, Hashable, Sendable {
    let planVersionID: UUID
    let programID: UUID
    let installedDayIDs: [UUID]

    enum CodingKeys: String, CodingKey {
        case planVersionID = "plan_version_id"
        case programID = "program_id"
        case installedDayIDs = "installed_day_ids"
    }
}
